byterun/io.c

```diff
@@ -78,6 +86,12 @@ CAMLexport struct channel * caml_open_descriptor_in(int fd)
   channel->revealed = 0;
   channel->old_revealed = 0;
   channel->refcount = 0;
+#ifdef _WIN32
+  h = (HANDLE) _get_osfhandle(fd);
+  if (GetFileType(h) == FILE_TYPE_CHAR && GetConsoleMode(h, &mode))
+    channel->flags = CHANNEL_FLAG_CONSOLE;
+  else
+#endif
   channel->flags = 0;
```

Comment:
I'm not fond of the way the windows-specific logic pollutes the function code (I think it breaks abstraction levels). Also, having an `else` at the end of an optional block is evil. Could you encapsulate this whole logic in a separate function?

(You could either have an ifdef _WIN32 here with `channel->flags = ...` in one case and `channel->flags = 0;` in the other, or have the function declared to always return 0 outside Windows, no strong preference here.)

Reply:
Meh - OK, I agree that ending the block with `else` was perhaps a little too towards the dark side.

I'm not sure that this breaks abstraction levels - handed an fd, the job of the function is to initialise the channel, which is what it's doing... it's a slightly unfair advantage to say that Unix can maintain the abstraction at an fd level, given that both C and the C runtime were invented for Unix's benefit...!

In this particular instance, I'm not certain about using a function, because it makes it tempting to use it elsewhere and the logic could become less clear. For example, one might be tempted to call this something caml_sys_isatty, but then #1406 might inadvertently affect it. I guess I could make it `channel->flags = initialise_channel_flags(fd);`, but I'm not that convinced by single-use functions.

Reply:
I apologize for being unclear; what I meant here and above is not that you are breaking abstraction boundaries, but that I think you are not following the "don't mix different levels of abstraction" principle (incidentally, it's the only idea in Clean Code-style books that I found both reasonable and not completely obvious). What this function was doing before was very simple and I could grok it in one pass: it's initializing each flag with a value that sort of looks right, adds it in a linked list (this could be a function call imho) and returns it. Now the code is mixed with non-trivial Win32-only logic, with two `ifdef` block that are to be understood together (as the variables bound in one are used in the other), I find the code much harder to read.